### DISCUSSION TITLE
[help] error:00000001:lib(0):func(0):reason(1), on the same machine

### DISCUSSION THREAD

**Discussion by xjj210130** [openssl/openssl#21900] [Q&A] [answered]
Hi all:
   I build the openssl version is 1.0.2k
    my env is centos 5.15.71. i use client and server on the same machine. but   there is the following wrong  SSL_accept error: error:00000001:lib(0):func(0):reason(1)

the code is following

```
int clientSocket = accept(sockHandle, (struct sockaddr*)(&client_addr), &client_addr_len);
SSL *ssl_accept = NULL;
nFlags = nFlags & (~O_NONBLOCK);
 if (fcntl(clientSocket, F_SETFL, nFlags) < 0)
            break;
ssl_accept = SSL_new(ctx_s);
if (0 == SSL_set_fd(ssl_accept, clientSocket)){}
SSL_set_accept_state(ssl_accept);
int k = SSL_accept(ssl_accept);
if(k < 1)
{
   printf("SSL_accept error: %s\n", error_str);
}
```

in fact the SSL_accept return code is less 1.
how to fix it. Thanks very much

**Comment by mattcaswell** [comment]
How did you get `error_str` in the above code?

**Comment by xjj210130** [comment]
yes, get the code is following:

 const char *error_str = ERR_error_string(eret, NULL);
  printf("SSL_accept error: %s\n", error_str);

the out is :
error:00000001:lib(0):func(0):reason(1), on the same machine

**Comment by mattcaswell** [comment]
An where did `eret` come from in the above code?

**Reply by xjj210130** [reply]
int clientSocket = accept(sockHandle, (struct sockaddr*)(&client_addr), &client_addr_len);
SSL *ssl_accept = NULL;
nFlags = nFlags & (~O_NONBLOCK);
 if (fcntl(clientSocket, F_SETFL, nFlags) < 0)
            break;
ssl_accept = SSL_new(ctx_s);
if (0 == SSL_set_fd(ssl_accept, clientSocket)){}
SSL_set_accept_state(ssl_accept);
int k = SSL_accept(ssl_accept);
if(k < 1)
{
    **int eret = SSL_get_error(ssl_accept, k);
    if (eret == SSL_ERROR_SYSCALL)
             ERR_clear_error(); /* emptie
   const char *error_str = ERR_error_string(eret, NULL);
   printf("SSL_accept error: %s\n", error_str);
   printf("SSL_accept error: %s\n", error_str);**
}

**Reply by xjj210130** [reply]
The para eret from  k<1.  Thanks very much

**Reply by t8m** [reply, marked as answer]
This is wrong use of ERR_error_string(). You need to use it on the value returned from one of the ERR_get_error(), ERR_peek_error(), or ERR_peek_last_error() functions. See https://www.openssl.org/docs/manmaster/man3/ERR_get_error.html

Please also see the https://www.openssl.org/docs/manmaster/man3/SSL_get_error.html manual page.

**Comment by xjj210130** [comment]
I use the following code:

```
int clientSocket = accept(sockHandle, (struct sockaddr*)(&client_addr), &client_addr_len);
SSL *ssl_accept = NULL;
nFlags = nFlags & (~O_NONBLOCK);
if (fcntl(clientSocket, F_SETFL, nFlags) < 0)
break;
ssl_accept = SSL_new(ctx_s);
if (0 == SSL_set_fd(ssl_accept, clientSocket)){}
SSL_set_accept_state(ssl_accept);
int k = SSL_accept(ssl_accept);
if(k < 1)
{
int eret = SSL_get_error(ssl_accept, k);
if (eret == SSL_ERROR_SYSCALL)
ERR_clear_error(); / emptie
const char error_str = ERR_error_string(eret, NULL);
printf("SSL_accept error: %s\n", error_str);

 int err = ERR_get_error();
 const char *error_str1 = ERR_error_string(err, NULL);
 printf("SSL_accept error_end: %s\n", error_str1);
 SSL_free(ssl_accept);
 close(clientSocket);
}

```

now the wrong message is :  error:1407609C:SSL routines:SSL23_GET_CLIENT_HELLO:http request

however , there is SSL_accept wrong.   In fact i send the data use SSL_write.  Both client and server  are c program

**Comment by mattcaswell** [comment]
> now the wrong message is : error:1407609C:SSL routines:SSL23_GET_CLIENT_HELLO:http request

This means that the client did not attempt a TLS handshake. Instead it sent an unsecured HTTP request.

**Comment by xjj210130** [comment]
In fact, I run the program , and test it .  It run about 10 minutes, There is no wrong. It run right.
However,  10 minutes  later, SSL_accept()  out the following message
00000000:lib(0):func(0):reason(0).
00000001:lib(0):func(0):reason(1)

 I am not familiar with openssl， Thanks very much
